aioxmpp/avatar/service.py

```diff
@@ -0,0 +1,536 @@
+        """
+        return self._remote_jid
+
+    @property
+    def url(self):
+        """
+        The URL where the avatar image data can be found.
+
+        This is :data:`None` if :attr:`image_in_pubsub` is true.
+        """
+        return self._url
+
+    @property
+    def width(self):
+        """
+        The width of the avatar image in pixels.
+
+        This is :data:`None` if this information is not supplied.
+        """
+        return self._width
+
+    @property
+    def height(self):
+        """
+        The height of the avatar image in pixels.
+
+        This is :data:`None` if this information is not supplied.
+        """
+        return self._height
+
+    @property
+    def nbytes(self):
+        """
+        The size of the avatar image data in bytes.
+        """
+        return self._nbytes
+
+    @property
+    def id_(self):
+        """
+        The SHA1 of the image data.
+
+        This is supposed to be used for caching.
+        """
+        return self._id
+
+    @property
+    def mime_type(self):
+        """
+        The MIME type of the image data.
+        """
+        return self._mime_type
+
+
+class PubsubAvatarDescriptor(AbstractAvatarDescriptor):
+
+    @property
+    def image_in_pubsub(self):
+        return True
+
+    @asyncio.coroutine
+    def get_image_bytes(self):
+        image_data = yield from self._pubsub.get_items_by_id(
+            self._remote_jid,
+            namespaces.xep0084_data,
+            [binascii.b2a_hex(self.id_).decode("ascii")],
+        )
+        if not image_data.payload.items:
+            raise RuntimeError("Avatar image data is not set.")
+
+        item, = image_data.payload.items
+        return item.registered_payload.data
+
+
+class HttpAvatarDescriptor(AbstractAvatarDescriptor):
+
+    @property
+    def image_in_pubsub(self):
+        return False
+
+    @asyncio.coroutine
+    def get_image_bytes(self):
+        """
+        Try to retrieve the avatar image date.
+
+        May raise NotImplementedError
+        """
+        raise NotImplementedError
+        yield
+
+
+class AvatarClient(service.Service):
+    """
+    Retrieves and caches avatar information of other clients. Emits
+    notifications on metadata changes.
+
+    .. autosignal:: on_metadata_changed
+
+    .. automethod:: get_avatar_metadata
+
+    .. automethod:: subscribe
+
+    """
+
+    ORDER_AFTER = [
+        disco.DiscoServer,
+        pubsub.PubSubClient,
+    ]
+
+    metadata_notify = disco.register_feature(
+        # register the notify capability for receiving avatar notifications
+        # automatically
+        # XXX: a PEPService on top of PubSubClient should handle this
+        # in the future
+        namespaces.xep0084_metadata + "+notify"
+    )
+
+    on_metadata_changed = callbacks.Signal(doc=
+    """
+    Fires when avatar metadata changes.
+
+    :param jid: The JID which the avatar belongs to.
+    :param metadata: The list of metadata descriptors.
+    """)
```

Comment:
Fire with ``# NOQA`` at this until the warnings are gone *or* put it inline into the class docstring with:

```
.. signal:: on_metadata_changed(…)
```

At the very least, you need to add the arguments to the ``.. autosignal::``, because they cannot be inferred from the signal declaration.